agent/consul/streaming_endpoint.go

```diff
@@ -26,34 +27,62 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
 		return fmt.Errorf("only ServiceHealth topic is supported")
 	}
 
-	snapshotCh := make(chan stream.Event, 32)
-	go snapshotFunc(server.Context(), snapshotCh, req.Key)
-
 	// Resolve the token and create the ACL filter.
 	rule, err := h.srv.ResolveToken(req.Token)
 	if err != nil {
 		return err
 	}
-	filt := newACLFilter(rule, h.srv.logger, h.srv.config.ACLEnforceVersion8)
+	aclFilter := newACLFilter(rule, h.srv.logger, h.srv.config.ACLEnforceVersion8)
 
-	// Wait for the events to come in and forward them to the client.
-	for event := range snapshotCh {
-		event.SetACLRules()
-		if !filt.allowEvent(event) {
-			continue
-		}
-		if err := server.Send(&event); err != nil {
-			return err
+	// Create the boolean expression filter.
+	var eval *bexpr.Evaluator
+	if req.Filter != "" {
+		eval, err = bexpr.CreateEvaluator(req.Filter, nil)
+		if err != nil {
+			return fmt.Errorf("Failed to create boolean expression evaluator: %v", err)
 		}
 	}
 
-	// Send a marker that this is the end of the snapshot.
-	endSnapshotEvent := stream.Event{
-		Topic:   req.Topic,
-		Payload: &stream.Event_EndOfSnapshot{EndOfSnapshot: true},
-	}
-	if err := server.Send(&endSnapshotEvent); err != nil {
-		return err
+	// Send an initial snapshot of the state via events if the requested index
+	// is lower than the last sent index of the topic.
+	lastSentIndex := state.LastTopicIndex(req.Topic)
+	if req.Index < lastSentIndex {
```

Comment:
What if `lastSentIndex` is `0`?

Reply:
If lastSentIndex is 0 then there's never been any activity on the topic and there shouldn't be any events to send, right?

Reply:
No, because this map is not persisted in the snapshot right now.

So on a new agent startup lastSendIndex will be zero since this agent has not yet sent any messages, but there may well be "events" i.e. service registered.

We'd need to persist that last sent map if you want to rely on that but I don't think it's necessary for now as long as we handle this case correctly (always send snap if the agent just restarted and doesn't know what it's sent before etc.)